### PR TITLE
changed adjoint data structure 

### LINKED ISSUE #22978: change adjoint data structure in OptimizationReporter
## Reason
The current adjoint data structure in the optimization reporter can only hold the derivative of the objective function with respect to a single type of variable.  In some problems, we will need derivatives of the objective function wrt multiple controllable types of parameters.  For instance, we may need to optimize boundary conditions on two different sidesets.

## Design
Change the adjoint data type to a vector of vectors.

### PR DESCRIPTION
**Pull request by lynnmunday**
use vector of vectors to hold gradients for multiple types of parameters closes #22978

### CODE REVIEW

**Comment by lynnmunday** [comment]
@zachmprince Will you look at this. For parameters = 'p1 p2' this creates a reporter called 'grad_wrt_p1 grad_wrt_p2'. I don't yet have a test where I transfer into the two different adjoint reporters, that test is still in isopod.  The tests we do have do properly create these two separate gradient reporters.

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C [hunk 1]
```diff
@@ -12,6 +12,31 @@
...
+  dof_id_type n = 0;
+  for (const auto & data : reporterVectors)
+    for (const auto & val : *data)
+      x.set(n++, val);
+
+  x.close();
+}
+
+void
+copyPetscVectorIntoReporter(const libMesh::PetscVector<Number> & x,
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/OptimizationReporter.C:30]: Make these static member functions in the OptimizationReporter class. Might be useful for other optimizations, so make them public.

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C [hunk 2]
```diff
@@ -64,10 +88,13 @@ OptimizationReporter::OptimizationReporter(const InputParameters & parameters)
                "Both upper and lower bounds must be specified if bounds are used");
 
   _parameters.reserve(_nparam);
+  _gradients.reserve(_nparam);
   unsigned int v = 0;
 
   for (const auto i : index_range(_parameter_names))
   {
+    _gradients.push_back(&declareValueByName<std::vector<Real>>("grad_wrt_" + _parameter_names[i],
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/OptimizationReporter.C:96]: I'd prefer without the `wrt`, but not a strong opinion.

File: modules/optimization/test/tests/optimizationreporter/base/gold/optRep_fromCsv_out.json [hunk 3]
```diff
@@ -1,173 +1,183 @@
 {
-  "reporters": {
```
Comment (zachmprince) [inline modules/optimization/test/tests/optimizationreporter/base/gold/optRep_fromCsv_out.json:2]: Why are these json files changing so much??? The names of the gradient vectors should be the only thing that's changing.

File: modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/main.i [hunk 4]
```diff
@@ -32,29 +32,14 @@
   #                       0.2 1.4 0
   #                       0.8 1.8 0'
   # measurement_values = '207 204 185 125'
+  outputs=none
```
Comment (zachmprince) [inline modules/optimization/test/tests/optimizationreporter/bc_load_linearFunction/main.i:35]: Why this?

**Review by zachmprince — CHANGES_REQUESTED** [review]

**Comment by lynnmunday** [comment]
@zachmprince will you review this again.  I started started over to get it to merge with your mesh reporter.

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C [hunk 5]
```diff
@@ -9,17 +9,70 @@
...
+copyPetscVectorIntoReporter(const libMesh::PetscVector<Number> & x,
+                            std::vector<std::vector<Real> *> reporterVectors)
+{
+  dof_id_type n = 0;
+  for (auto & data : reporterVectors)
+    for (auto & val : *data)
+      val = x(n++);
+}
+}
+
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:36]: I'd prefer to see these as public static functions of `OptimizationReporterBase`

File: modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C [hunk 6]
```diff
@@ -33,40 +31,33 @@ ParameterMeshOptimization::validParams()
...
-    _parameter_names(getParam<std::vector<ReporterValueName>>("parameter_names")),
-    _nparams(_parameter_names.size()),
-    _nvalues(_nparams),
-    _parameters(_nparams),
-    _gradients(_nparams),
-    _lower_bounds(getParam<std::vector<Real>>("lower_bounds")),
-    _upper_bounds(getParam<std::vector<Real>>("upper_bounds"))
+  : OptimizationReporterBase(parameters)
 {
+  _nvalues.resize(_nparams, 0); // fixme lynn this is different from optimization reporter
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C:59]: I don't think the fixme comment is necessary.

**Review by zachmprince — CHANGES_REQUESTED** [review]
Just a couple of minor suggestions.

**Comment by lynnmunday** [comment]
@zachmprince will you review this again

**Review by zachmprince — APPROVED** [review]

**Pull request merged**